### PR TITLE
(PUP-11723) Remove concurrent-ruby private class

### PR DESCRIPTION
**Pull request by mhashizume**
Prior to this commit, Puppet relied on the
Concurrent::RubyThreadLocalVar private class in concurrent-ruby.

This private class was removed in concurrent-ruby v1.2.0: https://github.com/ruby-concurrency/concurrent-ruby/commit/30465396

This commit updates Puppet::ThreadLocal to use
Concurrent::ThreadLocalVar instead of Concurrent::RubyThreadLocalVar and updates Puppet's dependency on concurrent-ruby to remove the pin to < v1.2.0.

### CODE REVIEW

**Comment by mhashizume** [comment]
Marking as draft as tests run

File: lib/puppet/thread_local.rb
Comment (joshcooper) [inline lib/puppet/thread_local.rb:5]: nit for future us, maybe update this text as we're now calling the JRuby implmentation on JRuby, not the Ruby one?

File: ext/project_data.yaml
Comment (Sharpie) [inline ext/project_data.yaml:28]: Seems like this would have to be `~> 1.2` as versions before that still created `java.lang.ThreadLocal` objects that became GC roots that would anchor JRuby interpreters and keep them in memory after Puppet Server tried to replace one of its pool workers.

**Review by Sharpie — CHANGES_REQUESTED** [review]

File: ext/project_data.yaml
Comment (joshcooper) [reply ext/project_data.yaml:28]: I was under the impression the leak was only when running in multithreaded mode. If that's not the case, then do you know if it was fixed upstream (when using the JRuby thread local var implementation on JRuby)?

Comment (joshcooper) [reply ext/project_data.yaml:28]: Copying a comment from slack. If we do need to require 1.2.0 or greater, then we'll need to bump the gem in the puppet-runtime and possible puppetserver. Otherwise, puppet will fail to load.

Comment (Sharpie) [reply ext/project_data.yaml:28]: From a quick skim of the Puppet source looking at uses of `Puppet::ThreadLocal`, it appears that we are using these objects in a few places without regard to whether multithreaded is enabled or not. Notably, `Puppet::Pops::PuppetStack` which holds the evaluation stack of the compiler.

I think the aim was to make Puppet thread-safe by default so that multithreaded could be switched on, not to make it thread-safe only when multithreaded is on.

Comment (mhashizume) [reply ext/project_data.yaml:28]: Thanks for the review @Sharpie . With the steps you provided over Slack, @joshcooper and I did some tests with the new code and older version of concurrent-ruby (1.1.10). It does not appear that this introduces a regression in the GC behavior, so I think we're good to move forward with this.

**Review by joshcooper — APPROVED** [review]

File: ext/project_data.yaml
Comment (justinstoller) [reply ext/project_data.yaml:28]: For posterity, what were the the steps?

Comment (mhashizume) [reply ext/project_data.yaml:28]: Good call @justinstoller. @Sharpie recommended:

1. Trigger catalog compilation to exercise `ThreadLocalVar`
2. Flush the JRuby pool via the API: `curl -X DELETE -kv --cert "$(puppet config print hostcert)" --key "$(puppet config print hostprivkey)" https://localhost:8140/puppet-admin-api/v1/jruby-pool`
3. Watch `puppetserver.log` to confirm how many JRuby interpreters are created/active, then run `jmap` to verify that there's not a discrepancy between between the two.

Comment (justinstoller) [reply ext/project_data.yaml:28]: Thanks.

I don't understand how 1.1.10 doesn't leak, but if you've confirmed the behavior no longer happens, maybe JRuby is now doing something to clean it up.

Comment (mhashizume) [reply ext/project_data.yaml:28]: @Sharpie can I get another review on this when you have a chance so we can merge? Thank you!

**Comment by joshcooper** [comment]
I'm going to merge this, but @Sharpie let us know if you have concerns.

**Pull request merged**
